shared/trieutil/sparse_merkle.go

```diff
@@ -10,64 +10,109 @@ import (
 )
 
-// MerkleTrie implements a sparse, general purpose Merkle trie to be used
+// SparseMerkleTrie implements a sparse, general purpose Merkle trie to be used
 // across ETH2.0 Phase 0 functionality.
-type MerkleTrie struct {
+type SparseMerkleTrie struct {
 	depth         uint
 	branches      [][][]byte
 	originalItems [][]byte // list of provided items before hashing them into leaves.
 }
 
 // NewTrie returns a new merkle trie filled with zerohashes to use.
-func NewTrie(depth int) (*MerkleTrie, error) {
+func NewTrie(depth int) (*SparseMerkleTrie, error) {
 	var zeroBytes [32]byte
 	items := [][]byte{zeroBytes[:]}
 	return GenerateTrieFromItems(items, depth)
 }
 
-// InsertIntoTrie inserts an item(deposit hash) into the trie.
-func (m *MerkleTrie) InsertIntoTrie(item []byte, index int) error {
-	// Only insert new items which follow directly after the last
-	// added element
-	if index > len(m.originalItems) {
-		return errors.New("invalid index to be inserting")
-	}
-	if index == len(m.originalItems) {
-		m.originalItems = append(m.originalItems, item)
-		return m.updateTrie()
-	}
-
-	m.originalItems[index] = item
-	return m.updateTrie()
-}
-
 // GenerateTrieFromItems constructs a Merkle trie from a sequence of byte slices.
-func GenerateTrieFromItems(items [][]byte, depth int) (*MerkleTrie, error) {
+func GenerateTrieFromItems(items [][]byte, depth int) (*SparseMerkleTrie, error) {
 	if len(items) == 0 {
 		return nil, errors.New("no items provided to generate Merkle trie")
 	}
-	layers := calcTreeFromLeaves(items, depth)
-	return &MerkleTrie{
+	leaves := items
+	layers := make([][][]byte, depth+1)
+	transformedLeaves := make([][]byte, len(leaves))
+	for i := range leaves {
+		arr := bytesutil.ToBytes32(leaves[i])
+		transformedLeaves[i] = arr[:]
+	}
+	layers[0] = transformedLeaves
+	for i := 0; i < depth; i++ {
+		if len(layers[i])%2 == 1 {
+			layers[i] = append(layers[i], zeroHashes[i])
+		}
+		updatedValues := make([][]byte, 0, 0)
+		for j := 0; j < len(layers[i]); j += 2 {
+			concat := hashutil.Hash(append(layers[i][j], layers[i][j+1]...))
+			updatedValues = append(updatedValues, concat[:])
+		}
+		layers[i+1] = updatedValues
+	}
+	return &SparseMerkleTrie{
 		branches:      layers,
 		originalItems: items,
 		depth:         uint(depth),
 	}, nil
 }
 
 // Items returns the original items passed in when creating the Merkle trie.
-func (m *MerkleTrie) Items() [][]byte {
+func (m *SparseMerkleTrie) Items() [][]byte {
 	return m.originalItems
 }
 
 // Root returns the top-most, Merkle root of the trie.
-func (m *MerkleTrie) Root() [32]byte {
+func (m *SparseMerkleTrie) Root() [32]byte {
 	enc := [32]byte{}
 	binary.LittleEndian.PutUint64(enc[:], uint64(len(m.originalItems)))
 	return hashutil.Hash(append(m.branches[len(m.branches)-1][0], enc[:]...))
 }
 
+// InsertIntoTrie inserts an item(deposit hash) into the trie.
+func (m *SparseMerkleTrie) InsertIntoTrie(item []byte, index int) {
+	for index >= len(m.branches[0]) {
```

Comment:
Given this is a sparse merkle trie, sometimes we may be attempting to insert something at index 15 when only 5 original leaves exist in the trie, so we have to expand it on-demand